crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,371 @@
+use std::{
+    alloc::{handle_alloc_error, Layout},
+    cell::UnsafeCell,
+    ptr::NonNull,
+};
+
+#[derive(Debug)]
+pub struct BlobVec {
+    item_layout: Layout,
+    capacity: usize,
+    len: usize,
+    data: UnsafeCell<NonNull<u8>>,
+    swap_scratch: UnsafeCell<NonNull<u8>>,
+    drop: unsafe fn(*mut u8),
+}
+
+impl BlobVec {
+    pub fn new(item_layout: Layout, drop: unsafe fn(*mut u8), capacity: usize) -> BlobVec {
+        if item_layout.size() == 0 {
+            BlobVec {
+                swap_scratch: UnsafeCell::new(NonNull::dangling()),
+                data: UnsafeCell::new(NonNull::dangling()),
+                capacity: usize::MAX,
+                len: 0,
+                item_layout,
+                drop,
+            }
+        } else {
+            let swap_scratch = NonNull::new(unsafe { std::alloc::alloc(item_layout) })
+                .unwrap_or_else(|| std::alloc::handle_alloc_error(item_layout));
+            let mut blob_vec = BlobVec {
+                swap_scratch: UnsafeCell::new(swap_scratch),
+                data: UnsafeCell::new(NonNull::dangling()),
+                capacity: 0,
+                len: 0,
+                item_layout,
+                drop,
+            };
+            blob_vec.reserve(capacity);
+            blob_vec
+        }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    pub fn reserve(&mut self, amount: usize) {
+        let available_space = self.capacity - self.len;
+        if available_space < amount {
+            self.grow(amount - available_space);
+        }
+    }
+
+    fn grow(&mut self, increment: usize) {
+        debug_assert!(self.item_layout.size() != 0);
+
+        let new_capacity = self.capacity + increment;
+        unsafe {
+            let new_data = if self.capacity == 0 {
+                std::alloc::alloc(
+                    array_layout(&self.item_layout, new_capacity)
+                        .expect("array layout should be valid"),
+                )
+            } else {
+                std::alloc::realloc(
+                    self.get_ptr().as_ptr(),
+                    array_layout(&self.item_layout, self.capacity)
+                        .expect("array layout should be valid"),
+                    self.item_layout.size() * new_capacity,
```

Comment:
```suggestion
                    array_layout(&self.item_layout, new_capacity).expect("array layout should be valid).size(),
```

Reply:
Can you also extract `array_layout(&self.item_layout, new_capacity)` into a new variable. There are three uses of it after applying the previous suggestion.

Reply:
once the layouts are moved to locals a match might be nicer here too
```rust
match self.capacity {
  0 => std::alloc::alloc(new_layout),
  _ => std::alloc::realloc(self.get_ptr().as_ptr(), old_layout, new_layout.size())
}